pkg/controller/federatedcluster/controller.go

```diff
@@ -160,82 +150,127 @@ func (cc *ClusterController) Run(stopChan <-chan struct{}) {
 		if err := cc.updateClusterStatus(); err != nil {
 			glog.Errorf("Error monitoring cluster status: %v", err)
 		}
-	}, cc.clusterMonitorPeriod, stopChan)
+	}, cc.clusterHealthCheckConfig.PeriodSeconds, stopChan)
 }
 
-// updateClusterStatus checks cluster status and get the metrics from cluster's restapi
+// updateClusterStatus checks cluster health and updates status of all FederatedClusters
 func (cc *ClusterController) updateClusterStatus() error {
 	clusters := &fedv1a1.FederatedClusterList{}
 	err := cc.client.List(context.TODO(), clusters, cc.fedNamespace)
 	if err != nil {
 		return err
 	}
 
-	for _, cluster := range clusters.Items {
+	var wg sync.WaitGroup
+	for _, obj := range clusters.Items {
 		cc.mu.RLock()
-		// skip updating status of the cluster which is not yet added to knownClusterSet.
-		if !cc.knownClusterSet.Has(cluster.Name) {
-			cc.mu.RUnlock()
-			continue
-		}
-		clusterClient, clientFound := cc.clusterKubeClientMap[cluster.Name]
-		clusterStatusOld, statusFound := cc.clusterStatusMap[cluster.Name]
+		cluster := obj.DeepCopy()
+		clusterStore := cc.clusterStoreMap[cluster.Name]
 		cc.mu.RUnlock()
-
-		if !clientFound {
-			glog.Warningf("Failed to get client for cluster %s", cluster.Name)
+		if clusterStore == nil {
+			glog.Warningf("Failed to retrieve stored data for cluster %s", cluster.Name)
 			continue
 		}
-		clusterStatusNew := clusterClient.GetClusterHealthStatus()
-		if !statusFound {
-			glog.Infof("There is no status stored for cluster: %v before", cluster.Name)
+
+		wg.Add(1)
+		go updateIndividualClusterStatus(cc.client, cluster, clusterStore, &cc.clusterHealthCheckConfig, &wg)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func updateIndividualClusterStatus(client genericclient.Client, cluster *fedv1a1.FederatedCluster,
+	store *ClusterStore, clusterHealthCheckConfig *util.ClusterHealthCheckConfig, wg *sync.WaitGroup) {
+	clusterClient := store.clusterKubeClient
+
+	clusterStatusNew := clusterClient.GetClusterHealthStatus()
+
+	if store.clusterStatus != nil {
+		if clusterStatusEqual(clusterStatusNew, store.clusterStatus) {
+			store.resultRun++
+			setLastTransitionTime(clusterStatusNew, store.clusterStatus.Conditions[0].LastTransitionTime)
 		} else {
-			hasTransition := false
-			if len(clusterStatusNew.Conditions) != len(clusterStatusOld.Conditions) {
-				hasTransition = true
-			} else {
-				for i := 0; i < len(clusterStatusNew.Conditions); i++ {
-					if !(strings.EqualFold(string(clusterStatusNew.Conditions[i].Type), string(clusterStatusOld.Conditions[i].Type)) &&
-						strings.EqualFold(string(clusterStatusNew.Conditions[i].Status), string(clusterStatusOld.Conditions[i].Status))) {
-						hasTransition = true
-						break
-					}
-				}
-			}
+			store.resultRun = 1
```

Comment:
There seems to be something wrong with setting this counter to 1 here.
Shouldn't this be after checking wrt thresholds?

Reply:
oops, some goof-up happened. Fixed it now. Also have added tests to verify.